<a href="https://colab.research.google.com/github/joekibz/moringa-wk5/blob/main/Thursday_Project_Notebook_Algorithms_and_Complexity_%5BJoseph_Thiongo%5D.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Project Notebook: Algorithms and Complexity 



## 1. Defining the Question

In this project, we put everything together we've learnt about algorithms complexity. We will imagine that we own an online laptop store and want to build a way to answer a few different business questions about our inventory.

We will use the laptops.csv (https://bit.ly/3G32iYj) file as our inventory.

* **ID**: A unique identifier for the laptop.
* **Company**: The name of the company that produces the laptop.
* **Product**: The name of the laptop.
* **TypeName**: The type of laptop.
* **Inches**: The size of the screen in inches.
* **ScreenResolution**: The resolution of the screen.
* **CPU**: The laptop CPU.
* **RAM**: The amount of RAM in the laptop.
* **Memory**: The size of the hard drive.
* **GPU**: The graphics card name.
* **OpSys**: The name of the operating system.
* **Weight**: The laptop weight.
* **Price**: The price of the laptop.

Let's explore the data and print the first few rows before we start implementing a way to represent this data as the store inventory.

**Pre-requisite Tasks**

* Import the csv module.
* Read the laptops.csv file using the `csv` module.
* Assign the first row to a variable named header.
* Assign the remaining rows to a variable named rows.
* Print the value of header.
* Print the first five rows in rows.

In [25]:
# Import the csv module
import csv

# Read the laptops.csv file
with open('laptops.csv', encoding = "ISO-8859-1") as f:
    reader = csv.reader(f)
    rows = list(reader)
    header = rows[0]
    rows = rows[1:]
    
print(header)
for i in range(5):
    print(rows[i])

['', 'Company', 'Product', 'TypeName', 'Inches', 'ScreenResolution', 'Cpu', 'Ram', 'Memory', 'Gpu', 'OpSys', 'Weight', 'Price_euros']
['1', 'Apple', 'MacBook Pro', 'Ultrabook', '13.3', 'IPS Panel Retina Display 2560x1600', 'Intel Core i5 2.3GHz', '8GB', '128GB SSD', 'Intel Iris Plus Graphics 640', 'macOS', '1.37kg', '1339.69']
['2', 'Apple', 'Macbook Air', 'Ultrabook', '13.3', '1440x900', 'Intel Core i5 1.8GHz', '8GB', '128GB Flash Storage', 'Intel HD Graphics 6000', 'macOS', '1.34kg', '898.94']
['3', 'HP', '250 G6', 'Notebook', '15.6', 'Full HD 1920x1080', 'Intel Core i5 7200U 2.5GHz', '8GB', '256GB SSD', 'Intel HD Graphics 620', 'No OS', '1.86kg', '575.00']
['4', 'Apple', 'MacBook Pro', 'Ultrabook', '15.4', 'IPS Panel Retina Display 2880x1800', 'Intel Core i7 2.7GHz', '16GB', '512GB SSD', 'AMD Radeon Pro 455', 'macOS', '1.83kg', '2537.45']
['5', 'Apple', 'MacBook Pro', 'Ultrabook', '13.3', 'IPS Panel Retina Display 2560x1600', 'Intel Core i5 3.1GHz', '8GB', '256GB SSD', 'Intel Iris P

## 2. Inventory Class

The goal of this project is to create a class that represents our inventory. The methods in that class will implement the queries that we want to answer about our inventory. We will also preprocess that data to make those queries run faster.

Here are some queries that we will want to answer:

* Given a laptop id, find the corresponding data.
* Given an amount of money, find whether there are two laptops whose total price is that given amount.
* Identify all laptops whose price falls within a given budget.
* Let's start by implementing the constructor. It will take the name of the CSV file as argument and then read the rows contained in it. You can use the code from the previous screen to do so.

Let's refresh our memories on how to create a class.

We can create a Python class named SomeName that has two arguments `argument1`, `argument2` using the following syntax:

```
class SomeName():

    def __init__(self, argument1, argument2):
        self.argument1 = argument1
        self.argument2 = argument2
```

The `self` parameter is a default parameter that is always passed to any method of the class. It contains a reference to the class itself and can be used to store values. In this case, we store the values of the two provided arguments.

**Tasks**

* Create a class named Inventory.
* Define the constructor (__init__() method) with two arguments: self and csv_filename.
* Read the CSV file provided in csv_filename. We will assume that the encoding is UTF-8, so you don't need to worry about it.
* Assign the first row to self.header and the remaining rows to self.rows.
Convert the price of each row to an integer. The price is the last column.
* Test your class by creating an instance of Inventory using 'laptops.csv' as argument.
* Print the headers by printing the value of the header property.
* Using the len() function, print the number of rows. You should have 1303 rows.

In [26]:
###file var...
file='laptops.csv'


In [27]:
# Challenge 
# ---
#
class Inventory():                    # step 1
    
    def __init__(self, csv_filename): # step 2
        with open(csv_filename,encoding = "ISO-8859-1" ) as f: # step 3
            reader = csv.reader(f)
            rows = list(reader)
        self.header = rows[0]         # step 4
        self.rows = rows[1:]
        for row in self.rows:         # step 5 ..switched str cast to [float] instead of [int] to clear value_encode error
            row[-1] = float(row[-1])

# your code goes here
###########
##Test your class by creating an instance of Inventory using 'laptops.csv' as argument.
inv=Inventory(file)

##Print the headers by printing the value of the header property.
print(inv.header)

##Using the len() function, print the number of rows. You should have 1303 rows.
len(rows)

#for i in range(5):
#   print(inv.rows[i])

['', 'Company', 'Product', 'TypeName', 'Inches', 'ScreenResolution', 'Cpu', 'Ram', 'Memory', 'Gpu', 'OpSys', 'Weight', 'Price_euros']


1303

## 3. Finding a Laptop From the Id

Throughout this project, we will ask you to make several improvements to the Inventory class. We suggest that you create a new cell at the start of each screen and copy and paste the last version of the class and modify the later. This will help you keep track of the changes that you make.

The first thing that we will implement is a way to look up a laptop from a given identifier. In this way, when a customer comes to our store with a purchase slip, we can quickly identify the laptop to which it corresponds.

**laptop id search**

For this, we will write a function named `get_laptop_from_id()`. This function will take as argument the identifier of the laptop and return the full row of the laptop with that id.


**Tasks**

* Inside the Inventory class, create a method `get_laptop_from_id()` with two arguments: self and laptop_id.
* Using a `for` loop over `self.rows`, identify if there is a row with whose laptop id is the same as laptop_id.
* Return that row if it was found or None if no laptop has the given identifier.
* Test your class by creating an instance of Inventory using `'laptops.csv'` as argument.
* Call `get_laptop_from_id()` by giving '3362737' as argument and print the result. It should find a matching laptop.
* Call `get_laptop_from_id()` by giving '3362736' as argument and print the result. It should not find a laptop.

In [28]:
# Challenge 
# ---
# 
class Inventory():                    
    
    def __init__(self, csv_filename): # step 2
        with open(csv_filename,encoding = "ISO-8859-1" ) as f: # step 3
            reader = csv.reader(f)
            rows = list(reader)
        self.header = rows[0]         # step 4
        self.rows = rows[1:]
        for row in self.rows:         # step 5 
            row[-1] = float(row[-1])
            
    def get_laptop_from_id(self, laptop_id):   # step 1
    # your code goes here
  
    ##Using a for loop over self.rows, identify if there is a row with whose laptop id is the same as laptop_id.
      for row in self.rows:
        if laptop_id in row:
          print(row)

      

        

##Test your class by creating an instance of Inventory using 'laptops.csv' as argument.
inv=Inventory(file)

##Call get_laptop_from_id() by giving '1112' as argument and print the result. It should find a matching laptop.
inv.get_laptop_from_id('1112')

##Call get_laptop_from_id() by giving '3362736' as argument and print the result. It should not find a laptop.
print("get_laptop_from_id(3362736): ",inv.get_laptop_from_id('3362736'))




['1112', 'MSI', 'GS70 Stealth', 'Gaming', '17.3', 'Full HD 1920x1080', 'Intel Core i7 6700HQ 2.6GHz', '8GB', '256GB SSD', 'Nvidia GeForce GTX 965M', 'Windows 10', '2.6kg', 1599.0]
get_laptop_from_id(3362736):  None


## 4. Improving Id Lookups

The algorithm we've created requires us to look at every single row to find the one that we are looking for (or decide that such a row does not exist). This algorithm has time complexity O(R) where R is the number of rows.

We've learned, that we can solve this problem more efficiently by preprocessing the data. Using a set, we can check in constant time whether a given identifier exists. However, we don't just want to know if it exists, we also want to retrieve the remaining row information. Therefore, we will use a dictionary instead of a set. Dictionaries have the same fast lookup properties that sets have, but allow us to associate values to the keys.

In our dataset, we only have about 1,300 laptops, so it might seem unnecessary to improve the performance of this query. However, you have to imagine that this code could be used in situations where the inventory contains millions of rows. Also, if we perform a lot of queries, even on a small dataset, the slow query performance will start to add up. It might eventually become the bottleneck of the application.

The idea is proceprocess the data into a dictionary where the keys are the IDs and the values the rows. Then, we will use that dictionary in the `get_laptop_from_id()` method. We can do this by:

* Preprocess the data and create the dictionary in the `__init__()` method.
* Re-implement the get_laptop_from_id() method. We will do it as a new method so that we can compare the two.


**Tasks** 

* At the end of the `__init__()` method, assign an empty dictionary to self.id_to_row.
* Loop over all rows and assign that row to the dictionary. Use the row id (the first element in a row) as the key and the whole row as the value.
* Create a new method named `get_laptop_from_id_fast()` with arguments: self and laptop_id.
* Implemented it by:
Checking whether the given id is in self.id_to_row.
* If it is, then return the corresponding row. Otherwise, return None.
* Test your class by creating an instance of Inventory using 'laptops.csv' as argument.
* Call `get_laptop_from_id_fast()` by giving '3362737' as argument and print the result. It should find a matching laptop.
Call `get_laptop_from_id_fast()` by giving `'3362736'` as argument and print the result. It should not find a laptop.

In [29]:
# Your code goes here
# ---
# 
class Inventory():                    
    
    def __init__(self, csv_filename): # step 2
        with open(csv_filename,encoding = "ISO-8859-1" ) as f: # step 3
            reader = csv.reader(f)
            rows = list(reader)
        self.header = rows[0]         # step 4
        self.rows = rows[1:]
        for row in self.rows:         # step 5 
            row[-1] = float(row[-1])

        ##Tasks
        ##At the end of the __init__() method, assign an empty dictionary to self.id_to_row.
        self.id_to_row={}

        ##Loop over all rows and assign that row to the dictionary.
        for row in self.rows:        
            self.id_to_row[row[0]] = row[1:]
    
    def get_laptop_from_id(self, laptop_id):   # step 1
    # your code goes here
  
    ##Using a for loop over self.rows, identify if there is a row with whose laptop id is the same as laptop_id.
      for row in self.rows:
        if laptop_id in row:
          print(row)

    
    ##Create a new method named get_laptop_from_id_fast() with arguments: self and laptop_id.
    def get_laptop_from_id_fast(self, laptop_id):
      #Implemented it by: Checking whether the given id is in self.id_to_row.
      #If it is, then return the corresponding row. Otherwise, return None.
      for row in self.rows:
        if laptop_id in self.id_to_row:
          print(self.id_to_row[laptop_id])
          break

      

        

##Test your class by creating an instance of Inventory using 'laptops.csv' as argument.
inv=Inventory(file)
#print(inv.id_to_row)


##Call get_laptop_from_id() by giving '1112' as argument and print the result. It should find a matching laptop.
inv.get_laptop_from_id_fast('1112')

##Call get_laptop_from_id() by giving '3362736' as argument and print the result. It should not find a laptop.
print("Get_laptop_from_id(3362736): ",inv.get_laptop_from_id_fast('3362736'))



['MSI', 'GS70 Stealth', 'Gaming', '17.3', 'Full HD 1920x1080', 'Intel Core i7 6700HQ 2.6GHz', '8GB', '256GB SSD', 'Nvidia GeForce GTX 965M', 'Windows 10', '2.6kg', 1599.0]
Get_laptop_from_id(3362736):  None


## 5. Comparing the Performance

The `get_laptop_from_id()` method has time complexity O(R) where R is the number of rows. In contrast, the new implementation as time complexity O(1). It does so by using more memory to store the self.`id_to_row` dictionary and using a bit more time creating an instance (because it needs to create the dictionary).

Let's experiment to compare the performance of the two methods. The idea is to generate random IDs using the random module. Then, use both methods to lookup these same IDs. We will use the time module to measure the execution time of each lookup and, for each method, add all times together.

Recall that we can measure the execution of a function `func()` as follows:

```
import time
start = time.time()
func()
end = time.time()
elapsed = end - start
```

To generate a list of N random values between a and b using the random module like so:
```
import random
values = [random.randint(a, b) for _ in range N]
```

**Tasks**

* Import the `time` module.
* Import the `random` module.
* Generate a list named `ids` with 10,000 random values between "1000000" and "9999999" (this is the id range). Note the use of strings rather than integers. This is because the IDs in the CSV files are read a strings, not integers. You can generate these by generating integers and converting them to strings using the str() function.

* Create an instance of Inventory by giving `'laptops.csv'` as argument.
Initialize a variable named `total_time_no_dict` and set it to 0. This variable will aggregate the times of calling get_laptop_from_id().

* For each identifier in ids do:
    a) Assign the value of time.time() to a variable named start.
    b) Call the get_laptop_from_id() function on the current identifier.
    c) Assign the value of time.time() to a variable named end.
	d) Add the elapsed time, end - start, to total_time_no_dict.

* Initialize a variable named `total_time_dict` and set it to 0. This variable will aggregate the times of calling get_laptop_from_id_fast().
For each identifier in ids do:
	a) Assign the value of time.time() to a variable named start.
	b) Call the get_laptop_from_id_fast() function on the current identifier.
	c) Assign the value of time.time() to a variable named end.
	d) Add the elapsed time, end - start, to total_time_dict.

* Print the values of `total_time_no_dict` and `total_time_dict`.

In [30]:
# Your code goes here
import time
import random

##Generate a list named ids with 10,000 random values between "1000000" and "9999999" (this is the id range)
ids = [str(random.randint(1, 1303)) for _ in range (10000)]
#ids

##Create an instance of Inventory by giving 'laptops.csv' as argument.
inv=Inventory(file)

##get total time no dictionary 
def getTimeNoDict(ids, file):
  totalTime=0
  for i in ids:
    start=time.time()
    file.get_laptop_from_id(i)
    end=time.time()
    elapsed=end-start
    totalTime+=elapsed

  return totalTime

##get total time WITH dictionary 
def getTimeWithDict(ids, file):
  totalTime=0
  for i in ids:
    start=time.time()
    file.get_laptop_from_id_fast(i)
    end=time.time()
    elapsed=end-start
    totalTime+=elapsed

  return totalTime

##Initialize a variable named total_time_no_dict and set it to 0.
total_time_no_dict=getTimeNoDict(ids,inv)

##Initialize a variable named total_time_dict and set it to 0. This variable will aggregate the times of calling get_laptop_from_id_fast()
total_time_dict=getTimeWithDict(ids, inv)

Streaming output truncated to the last 5000 lines.
['Dell', 'Alienware 17', 'Gaming', '17.3', '2560x1440', 'Intel Core i7 6820HK 2.7GHz', '16GB', '512GB SSD +  1TB HDD', 'Nvidia GeForce GTX 1080', 'Windows 10', '4.36kg', 2758.0]
['Asus', 'VivoBook E403NA', 'Notebook', '14', 'Full HD 1920x1080', 'Intel Celeron Dual Core N3350 1.1GHz', '4GB', '32GB Flash Storage', 'Intel HD Graphics 500', 'Windows 10', '1.5kg', 298.0]
['Dell', 'Inspiron 3567', 'Notebook', '15.6', 'Full HD 1920x1080', 'Intel Core i5 7200U 2.5GHz', '4GB', '256GB SSD', 'AMD Radeon R5 M430', 'Linux', '2.3kg', 598.9]
['Dell', 'Alienware 17', 'Gaming', '17.3', 'IPS Panel Full HD 1920x1080', 'Intel Core i7 7700HQ 2.8GHz', '32GB', '256GB SSD +  1TB HDD', 'Nvidia GeForce GTX 1070', 'Windows 10', '4.42kg', 3149.0]
['Lenovo', 'Thinkpad X270', 'Ultrabook', '12.5', 'IPS Panel Full HD 1920x1080', 'Intel Core i7 7500U 2.7GHz', '16GB', '512GB SSD', 'Intel HD Graphics 620', 'Windows 10', '1.36kg', 1798.0]
['MSI', 'GS73VR 7RF', 'Gaming', 

```
# Success - Pre-processing using dictionary improves processing time !...
```

In [31]:
#Print the values of total_time_no_dict and total_time_dict.

print("Total time no dictionary: ",total_time_no_dict)

print("Total time using dictionary: ", total_time_dict)

Total time no dictionary:  9.995435953140259
Total time using dictionary:  3.960927724838257


## 6. Two Laptop Promotion

Sometimes, your store offers a promotion where you give a gift card. A customer can use the gift to buy up to two laptops. To avoid having to keep track of what was already spent, the gift card has a single time usage. This means that, even if there is leftover money, it cannot be used anymore.

For example, imagine that there are only three laptops in inventory. The prices of these laptops are $1,339, $898, and $575.

Say we offered a gift card of $2,500. 

Since a customer can buy, at most, two laptops with a gift card, the maximum they can spend is $2,237 ($1,339 plus $898).

Therefore, they might feel cheated because, no matter how they spend their gift card, they cannot spend the full $2,500.

You don't want to make a customer feel cheated, so whenever you issue a gift card, you want to make sure that there is at least one way to spend it in full. In other words, before issuing a gift card for D dollars, you want to make sure that either there is a laptop that costs exactly D dollars or two laptops whose costs add up to precisely D dollars.

Write a function that, given a dollar amount, checks whether it is possible to spend precisely that amount by purchasing up to two laptops.

**Tasks** 

* Create a method named `check_promotion_dollars()` that takes two arguments: self and dollars.
* Loop over all rows to check if there exists a laptop whose price is exactly dollars. Return True if you find one.
* Using a double for loop, iterate over all pairs of rows (not necessarily distinct because we can buy the same laptop twice) and check if there is a pair whose prices adds up to exactly dollars. Return True if you find one.
* At the end of the function, return `False` to indicate that it is impossible to spend exactly dollars by purchasing at most two laptops.
* Test your class by creating an instance of Inventory by giving `'laptops.csv'` as argument.
* Call `check_promotion_dollars()` by giving 1000 as argument and print the result. It should find a solution.
* Call `check_promotion_dollars()` by giving 442 as argument and print the result. It should not find a solution.


In [32]:
# Your code goes here

class Inventory():                    
    
    def __init__(self, csv_filename): # step 2
        with open(csv_filename,encoding = "ISO-8859-1" ) as f: # step 3
            reader = csv.reader(f)
            rows = list(reader)
        self.header = rows[0]         # step 4
        self.rows = rows[1:]
        for row in self.rows:         # step 5 
            row[-1] = float(row[-1])

        ##Tasks
        ##At the end of the __init__() method, assign an empty dictionary to self.id_to_row.
        self.id_to_row={}

        ##Loop over all rows and assign that row to the dictionary.
        for row in self.rows:        
            self.id_to_row[row[0]] = row[1:]
    
    def get_laptop_from_id(self, laptop_id):   # step 1
    # your code goes here
  
    ##Using a for loop over self.rows, identify if there is a row with whose laptop id is the same as laptop_id.
      for row in self.rows:
        if laptop_id in row:
          print(row)

    
    ##Create a new method named get_laptop_from_id_fast() with arguments: self and laptop_id.
    def get_laptop_from_id_fast(self, laptop_id):
      #Implemented it by: Checking whether the given id is in self.id_to_row.
      #If it is, then return the corresponding row. Otherwise, return None.
      for row in self.rows:
        if laptop_id in self.id_to_row:
          print(self.id_to_row[laptop_id])
          break

    ##Tasks ......................
    ##Create a method named check_promotion_dollars() that takes two arguments: self and dollars.
    def check_promotion_dollars(self, dollars):
      #Loop over all rows to check if there exists a laptop whose price is exactly dollars. Return True if you find one.
      for row in self.rows:
            if row[-1] == dollars:
              return True

      #Using a double for loop, iterate over all pairs of rows (not necessarily distinct because we can buy the same laptop twice) and check if there is a pair whose prices adds up to exactly dollars
      for i in self.rows:
        for j in self.rows:
            if i[-1] + j[-1] == dollars:
              return True
      
      #At the end of the function, return False to indicate that it is impossible to spend exactly dollars by purchasing at most two laptops
      return False

##Test your class by creating an instance of Inventory by giving 'laptops.csv' as argument
inv=Inventory('laptops.csv')

#Call check_promotion_dollars() by giving 1000 as argument and print the result. It should find a solution
print("1000USD promo possible?... ",inv.check_promotion_dollars(1000))

#Call check_promotion_dollars() by giving 442 as argument and print the result. It should not find a solution.
print("442USD promo possible?... ",inv.check_promotion_dollars(442))



1000USD promo possible?...  True
442USD promo possible?...  False


## 7. Optimizing Laptop Promotion

We've learned how we can preprocess data to answer the kind of queries that we used in the `check_promotion_dollars()`. Let's implement this to make our code run faster.

Since we only care about whether or not there is a solution, we can store all laptops prices in a set when we initialize the inventory. Then we can check in constant time whether there is a laptop with a given price.

To check if there is a pair of laptops can be done in the same way as we've learned in the last lesson.


**Tasks**
* At the end of the `__init__()` method, assign an empty set to self.prices.
* Loop over all rows and add the price contained in that row to `self.prices`.
* Create a method named `check_promotion_dollars_fast()` that takes two arguments: self and dollars.
* Use the `self.prices` set to check whether there is a laptop whose cost is exactly dollars. Return `True` if it is the case.
* Using the technique we've learned in the previous lesson to check whether two values in `self.prices` add up to exactly dollar. Return `True` if it is the case.
* At the end of the function, return False to indicate that it is impossible to spend exactly dollars by purchasing at most two laptops.
* Test your class by creating an instance of Inventory by giving `'laptops.csv'` as argument.
* Call `check_promotion_dollars_fast()` by giving 1000 as argument and print the result. It should find a solution.
* Call `check_promotion_dollars_fast()` by giving 442 as argument and print the result. It should not find a solution.

In [33]:
# Your code goes here

class Inventory():                    
    
    def __init__(self, csv_filename): # step 2
        with open(csv_filename,encoding = "ISO-8859-1" ) as f: # step 3
            reader = csv.reader(f)
            rows = list(reader)
        self.header = rows[0]         # step 4
        self.rows = rows[1:]
        for row in self.rows:         # step 5 
            row[-1] = float(row[-1])

        ##Tasks
        ##At the end of the __init__() method, assign an empty dictionary to self.id_to_row.
        self.id_to_row={}

        ##Loop over all rows and assign that row to the dictionary.
        for row in self.rows:        
            self.id_to_row[row[0]] = row[1:]

        ##Task7- At the end of the __init__() method, assign an empty set to self.prices.
        self.prices = set(())

        ##Task7- Loop over all rows and add the price contained in that row to self.prices.
        for row in self.rows:  
            self.prices.add(row[-1])
    
    def get_laptop_from_id(self, laptop_id):   # step 1
    # your code goes here
  
    ##Using a for loop over self.rows, identify if there is a row with whose laptop id is the same as laptop_id.
      for row in self.rows:
        if laptop_id in row:
          print(row)

    
    ##Create a new method named get_laptop_from_id_fast() with arguments: self and laptop_id.
    def get_laptop_from_id_fast(self, laptop_id):
      #Implemented it by: Checking whether the given id is in self.id_to_row.
      #If it is, then return the corresponding row. Otherwise, return None.
      for row in self.rows:
        if laptop_id in self.id_to_row:
          print(self.id_to_row[laptop_id])
          break

    
    ##Task6 - Create a method named check_promotion_dollars() that takes two arguments: self and dollars.
    def check_promotion_dollars(self, dollars):
      #Loop over all rows to check if there exists a laptop whose price is exactly dollars. Return True if you find one.
      for row in self.rows:
            if row[-1] == dollars:
              return True

      #Using a double for loop, iterate over all pairs of rows (not necessarily distinct because we can buy the same laptop twice) and check if there is a pair whose prices adds up to exactly dollars
      for i in self.rows:
        for j in self.rows:
            if i[-1] + j[-1] == dollars:
              return True
      
      #At the end of the function, return False to indicate that it is impossible to spend exactly dollars by purchasing at most two laptops
      return False

    ##Task7 - Create a method named check_promotion_dollars_fast() that takes two arguments: self and dollars
    def check_promotion_dollars_fast(self, dollars):
        #Use the self.prices set to check whether there is a laptop whose cost is exactly dollars. Return True if it is the case
        for i in self.prices:
          if i == dollars:
            return True 

        #Using the technique we've learned in the previous lesson to check whether two values in self.prices add up to exactly dollar. 
        #Return True if it is the case.
        for i in self.prices:
          for j in self.prices:
            if i + j == dollars:
              return True


        #At the end of the function, return False to indicate that it is impossible to spend exactly dollars by purchasing at most two laptops.
        return False

##Test your class by creating an instance of Inventory by giving 'laptops.csv' as argument.
inv=Inventory('laptops.csv')

##Call check_promotion_dollars_fast() by giving 1000 as argument and print the result. It should find a solution.
print("1000USD promo possible?... ",inv.check_promotion_dollars_fast(1000))

##Call check_promotion_dollars_fast() by giving 442 as argument and print the result. It should not find a solution.
print("442USD promo possible?... ",inv.check_promotion_dollars_fast(442))


1000USD promo possible?...  True
442USD promo possible?...  False


## 8. Comparing Promotion Functions

Let's compare the performance of the last two functions that we wrote.

* Generate a list named `prices` with 100 random values between 100 and 5,000.
* Create an instance of `Inventory` by giving `'laptops.csv'` as argument.
* Initialize a variable named `total_time_no_set` and set it to 0. This variable will aggregate the times of calling `check_promotion_dollars()`.
* For each value in the prices do:
    a) Assign the value of `time.time()` to a variable named `start`.
    b) Call the `check_promotion_dollars()` function on the current price.
    c) Assign the value of `time.time()` to a variable named `end`.
    d) Add the elapsed time, end - start, to `total_time_no_set`.
* Initialize a variable named `total_time_set` and set it to `0`. This variable will aggregate the times of calling `check_promotion_dollars_fast()`.
* For each value in the prices do:
	a) Assign the value of time.time() to a variable named start.
	b) Call the check_promotion_dollars_fast() function on the current price.
	c) Assign the value of `time.time()` to a variable named end.
	d) Add the elapsed time, end - start, to `total_time_set`.
* Print the values of `total_time_no_set` and `total_time_set`.

In [34]:
# Your code goes below
#Task8 - Generate a list named prices with 100 random values between 100 and 5,000.
prices = [random.randint(100, 5000) for _ in range (100)]

#Create an instance of Inventory by giving 'laptops.csv' as argument.
inv=Inventory('laptops.csv')

##get total time no set 
def getTimeNoSet(prices, file):
  totalTime=0
  for i in prices:
    start=time.time()
    file.check_promotion_dollars(i)
    end=time.time()
    elapsed=end-start
    totalTime+=elapsed

  return totalTime

##get total time WITH set 
def getTimeWithSet(prices, file):
  totalTime=0
  for i in prices:
    start=time.time()
    file.check_promotion_dollars_fast(i)
    end=time.time()
    elapsed=end-start
    totalTime+=elapsed

  return totalTime

##Initialize a variable named total_time_no_set.
total_time_no_set=getTimeNoSet(prices,inv)

##Initialize a variable named total_time_set 
total_time_set=getTimeWithSet(prices, inv)

```
# Success - Sets process faster than dictionaries !...
```



In [35]:
#Print time values result
print("Total time no set: ",total_time_no_set)

print("Total time using set: ", total_time_set)

Total time no set:  1.4585518836975098
Total time using set:  0.3755466938018799


## 9. Conclusion

Congratulations on implementing a class to represent the inventory of a laptop shop. In this project, you've learned that we can answer business questions more efficiently by preprocessing the data.


If you want to push this project further, we suggest that you think about the following queries:

* Imagine that we extend our budget query to take as input a range of prices, `min_price` and `max_price`, rather than a single price. Write a query that finds all laptops whose price is in the given range.
Sometimes, a customer wants a laptop with some characteristics such as, for instance, 8GB or RAM and a 256GB hard drive. It would be interesting for those customers to provide a way to find the cheapest laptop that matches the desired characteristics. For simplicity, focus only on the amount of RAM and hard drive capacity. You might need to convert those values to integers rather than using strings.
* In this project, we only explored three possible queries that we might want to do over the data. In general, we often have a lot of different datasets to process and queries to answer. Designing such a class for every type of data in a business and implementing specific query methods takes a lot of time
